lib/MIDI/RtMidi/FFI/ScorePlayer.pm

```diff
@@ -0,0 +1,141 @@
+        # and DLSMusicDevice open in GarageBand / Logic with this
+        # sound front selected.
+        # DLSMusicDevice should receive input from the virtual port
+        # opened below.
+        # See MIDI::RtMidi::FFI::Device docs for more info.
+        $opts{device}->open_virtual_port( 'foo' ) if $^O eq 'darwin';
+        # Alternatively you can use FluidSynth
+        $opts{device}->open_port_by_name( $opts{port} );
+
+        bless \%opts, $class;
+    }
+
+    sub device { shift->{ device } } # XXX Is this even needed?
+
+    # This manipulates internals of MIDI::Score objects and
+    # hashes used by drum-circle - doing this isn't a good
+    # idea - skip to `sub play` to see the interesting piece
+    # of this example.
+    sub _reset_score {
+        my ( $self ) = @_;
+        # sorry
+        $self->{score}->{ Score } = [
+            grep { $_->[0] !~ /^note/ }
+            @{ $self->{score}->{ Score } }
+        ];
+        ${ $self->{score}->{ Time } } = 0;
+        $self->{common}{seen} = {}
+            if exists $self->{common}{seen};
+    }
+
+    sub play {
+        my ( $self ) = @_;
+        while( 1 ) {
+            $self->_sync_phrases;
+            my $micros = get_microseconds($self->{score});
+            my $events = score2events($self->{score});
+            for my $event (@{ $events }) {
+                next if $event->[0] =~ /set_tempo|time_signature/;
+                if ( $event->[0] eq 'text_event' ) {
+                    printf "%s\n", $event->[-1];
+                    next;
+                }
+                my $useconds = $micros * $event->[1];
+                usleep($useconds) if ( $useconds > 0 && $useconds < 1_000_000 );
+                $self->device->send_event( $event->[0] => @{ $event }[ 2 .. $#$event ] );
+            }
+            sleep(2);
+            $self->_reset_score;
+        }
+    }
+
+    # Build the code-ref MIDI of all phrases to be played
+    sub _sync_phrases {
+        my ( $self ) = @_;
+        my @phrases;
+        my $n = 1;
+        push @phrases, $_->( %{ $self->{common} }, phrase => $n++ ) 
+            for @{ $self->{phrases} };
+        $self->{score}->synch( @phrases ); # Play the phrases simultaneously
+    }
+
+};
+
+1;
+
+__END__
+
+=head1 NAME
+
+MIDI::RtMidi::FFI::ScorePlayer
+
+=head1 SYNOPSIS
+
+  use MIDI::RtMidi::FFI::ScorePlayer ();
+  use MIDI::Util qw(setup_score);
+
+  my $score = setup_score();
+
+  my %common = ( score => $score, seen => {}, etc => '...', );
+
+  sub treble {
+      my ( %args ) = @_;
+      my $treble = sub {
+          ...; # Add notes or rests to the score
+      }
+      return $treble;
+  }
+  sub bass {
+      ...; # As above
+  }
+
+  ScorePlayer->new(
+      score   => $score,
+      phrases => [ \&treble, \&bass ],
+      common  => \%common,
+  )->play;
+
+=head1 DESCRIPTION
+
+TBD
+
+=head1 SEE ALSO
+
+TBD
+
+=head1 AUTHOR
+
+JBARRETT
+
+=head1 COPYRIGHT AND LICENSE
```

Comment:
You should make yourself the author / copyright holder - most of `sub play` was pulled from your code as well.

Not sure how PodWeaver / dzil will handle this but we'll find out :)